Add support for multiple CA pins

nklaassen opened the pull request:
HA clusters with multiple HSM auth servers will have different keys at each auth server. If the auth servers are behind a load balancer, we can't guarantee which auth server and which TLS cert will be used. Commands which use a CA pin should allow multiple CA pins to be specified so that the auth server will be guaranteed to match one of them.

This PR also adds the API to get the list of all the required CA pins and adds them to the output of `tctl status` and `tctl get token`. The buddy teleport.e PR https://github.com/gravitational/teleport.e/pull/303 adds all CA pins to the automatic join script.

nklaassen committed:
add support for multiple CA pins

xacrimon reviewed (approved):
lgtm

tcsc commented on lib/utils/spki.go:
```diff
@@ -33,26 +33,39 @@ func CalculateSPKI(cert *x509.Certificate) string {
 }
 
-var errorMessage string = "provided certificate pin does not match cluster pin. " +
+var errorMessage string = "cluster pin does not match any provided certificate pin. " +
```
```suggestion
var errorMessage string = "cluster pin does not match any provided certificate authority pin. " +
```

nklaassen committed:
Update error message

Co-authored-by: Trent Clarke <trent@goteleport.com>

nklaassen committed:
Merge branch 'master' into nklaassen/ca-pin-multiple-keys

r0mant commented on lib/auth/apiserver.go:
```diff
@@ -1137,14 +1138,33 @@ func (s *APIServer) getDomainName(auth ClientI, w http.ResponseWriter, r *http.R
+		return nil, trace.Wrap(err)
+	}
+
+	return LocalCAResponse{
```
Is this a backwards-compatible change? Seems like the format of the returned response is changing, no?

r0mant commented on tool/tctl/common/token_command.go:
```diff
@@ -288,16 +281,20 @@ func (c *TokenCommand) List(client auth.ClientI) error {
-	localCA, err := client.GetClusterCACert()
+// calculateCAPins returns the SPKI pins for the local cluster.
+func calculateCAPins(client auth.ClientI) ([]string, error) {
+	caCerts, err := client.GetClusterCACerts()
```
If the client is talking to an older auth server, this won't work. Can we do a fallback to the old API method here?

nklaassen replied:
I ended up updating the client `GetClusterCACerts` method to fall back to the old endpoint on error

nklaassen replied:
I've not also added back to old API method for compat with a slightly out of date `teleport.e` so I can get CI to pass. I tested this with both new client/old server and new server/old client, it works for both.

r0mant commented on lib/auth/clt.go:
```diff
@@ -406,17 +406,17 @@ func (c *Client) GetDomainName() (string, error) {
 }
 
-// GetClusterCACert returns the CAs for the local cluster without signing keys.
-func (c *Client) GetClusterCACert() (*LocalCAResponse, error) {
```
I think we might need to keep this method for compatibility with older auth servers (see my comment somewhere below).

nklaassen committed:
Merge branch 'master' into nklaassen/ca-pin-multiple-keys

nklaassen committed:
address review comments

nklaassen commented on lib/auth/auth.go:
```diff
@@ -480,13 +480,14 @@ func (a *Server) GetClusterCACert() (*LocalCAResponse, error) {
+
+	certs := services.GetTLSCerts(hostCA)
+	if len(certs) < 1 {
+		return nil, trace.BadParameter("no tls certs found in host CA")
```
yes, fixed

nklaassen commented on lib/auth/apiserver.go:
```diff
@@ -1137,14 +1138,33 @@ func (s *APIServer) getDomainName(auth ClientI, w http.ResponseWriter, r *http.R
+	}
+
+	return LocalCAResponse{
+		TLSCA: certs.Certs[0],
```
done

nklaassen committed:
avoid returning slice with nil value

nklaassen committed:
revert teleport.e ref to master's version

nklaassen committed:
re-introduce deprecated GetClusterCACert methods

needed for compat with slightly outdated teleport.e

r0mant commented on lib/auth/apiserver.go:
```diff
@@ -96,6 +96,7 @@ func NewAPIServer(config *APIConfig) (http.Handler, error) {
 	// Operations on certificate authorities
 	srv.GET("/:version/domain", srv.withAuth(srv.getDomainName))
 	srv.GET("/:version/cacert", srv.withAuth(srv.getClusterCACert))
+	srv.GET("/:version/cacerts", srv.withAuth(srv.getClusterCACerts))
```
Sorry I have only just realized now that you're adding this method to the HTTP API which is deprecated. All new API methods should be added to GRPC server IIRC. @Joerger Is this correct?

awly replied:
Yeah, let's add this to the gRPC service and make it either (1) return all CAs of all types or (2) return CAs of one type specified as a request parameter. Don't make it return `HostCA` only, implicitly.

nklaassen replied:
I just switched this to a new gRPC method but just saw your comment now @awly about potentially returning other CA certs. Do you have a preference between returning all certs or requiring the parameter? Both are doable

r0mant commented on lib/auth/clt.go:
```diff
@@ -419,6 +420,29 @@ func (c *Client) GetClusterCACert() (*LocalCAResponse, error) {
+func (c *Client) GetClusterCACerts() (*ClusterCACertsResponse, error) {
+	var caCerts ClusterCACertsResponse
+	out, err := c.Get(c.Endpoint("cacerts"), url.Values{})
+	if err != nil {
```
Can we check specifically for "not implemented" error here? That's what we do in other places IIRC if you add the new methods to GRPC server according to my other comment.

nklaassen replied:
Yes, that's easier with the gRPC method, it does that in the latest commit.

nklaassen commented:
Reverted to before I added the gRPC method as I forgot it's important that this endpoint does not require any authentication for new nodes to join. Reaching out to @r0mant and @awly offline about the best way to move forward here.

nklaassen committed:
single endpoint returns all concatenated certs

nklaassen committed:
update comments

nklaassen committed:
fix lint

nklaassen committed:
Merge branch 'master' into nklaassen/ca-pin-multiple-keys

nklaassen commented:
@r0mant @awly Andrew came up with the idea of just concatenating all of the PEM certs in the existing `cacert` response, so that's what I've done here. Should be compatible with older versions as they will only read the first PEM block (I've tested this with tctl, actual nodes trying to join with ca pins won't be able to without support for multiple pins anyways). I appreciate your guys' reviews on this.

awly commented on lib/auth/auth.go:
```diff
@@ -485,13 +487,20 @@ func (a *Server) GetClusterCACert() (*LocalCAResponse, error) {
+		return nil, trace.NotFound("no tls certs found in host CA")
+	}
+
+	allCerts := certs[0]
+	for i := 1; i < len(certs); i++ {
+		allCerts = append(allCerts, '\n')
+		allCerts = append(allCerts, certs[i]...)
 	}
```
is this a `bytes.Join(certs, []byte("\n"))`?

nklaassen replied:
Yep, thanks

awly commented on lib/auth/register.go:
```diff
@@ -315,33 +315,37 @@ func pinRegisterClient(params RegisterParams) (*Client, error) {
+		}
 
-	log.Infof("Joining remote cluster %v with CA pin.", tlsCA.Subject.CommonName)
+		log.Infof("Joining remote cluster %v with CA pin.", cert.Subject.CommonName)
+	}
```
```suggestion
	}
	log.Infof("Joining remote cluster %v with CA pin.", certs[0].Subject.CommonName)
```

awly commented on lib/auth/register.go:
```diff
@@ -315,33 +315,37 @@ func pinRegisterClient(params RegisterParams) (*Client, error) {
 	certPool := x509.NewCertPool()
-	certPool.AddCert(tlsCA)
+	for _, cert := range certs {
+		certPool.AddCert(cert)
```
check the return value

nklaassen replied:
There is no return value :)

awly commented on lib/config/fileconf.go:
```diff
@@ -350,9 +350,13 @@ type Global struct {
 	// If omitted, the default will be used.
 	CASignatureAlgorithm *string `yaml:"ca_signature_algo,omitempty"`
 
+	// DEPRECATED: Prefer CAPins
 	// CAPin is the SKPI hash of the CA used to verify the Auth Server.
 	CAPin string `yaml:"ca_pin"`
 
+	// CAPins is the list of SKPI hashes fo the CAs used to verify Auth Servers.
+	CAPins []string `yaml:"ca_pins"`
+
```
I think we have `apiutils.Strings`, which handles unmarshaling YAML as either a single string or a list of strings.
You can use that to keep the old field name.

nklaassen replied:
good point, switched that, nice to avoid a new field name here

zmb3 replied:
@awly @nklaassen I think this actually broke the case where no CA pin is provided, as `apiutils.Strings("")` unmarshals to `[]string{""}` instead of an empty slice.

awly commented on lib/utils/spki.go:
```diff
@@ -33,26 +33,39 @@ func CalculateSPKI(cert *x509.Certificate) string {
-		return trace.BadParameter(errorMessage)
+	// Timing of this check depends only on the number of pins and certs, not
+	// their contents.
+	for _, cert := range certs {
+		certHasMatchingPin := false
+		for _, pin := range pins {
+			// Check that that pin itself matches that value calculated from the passed
+			// in certificate.
+			if subtle.ConstantTimeCompare([]byte(CalculateSPKI(cert)), []byte(pin)) == 1 {
+				certHasMatchingPin = true
+			}
+		}
+		if !certHasMatchingPin {
+			return trace.BadParameter(errorMessage)
+		}
 	}
```
```suggestion
	outer:
	for _, cert := range certs {
		for _, pin := range pins {
			// Check that that pin itself matches that value calculated from the passed
			// in certificate.
			if subtle.ConstantTimeCompare([]byte(CalculateSPKI(cert)), []byte(pin)) == 1 {
				continue outer
			}
		}
		return trace.BadParameter(errorMessage)
	}
```

r0mant reviewed (approved):
lgtm once @awly's comments are addressed

nklaassen committed:
update log

Co-authored-by: Andrew Lytvynov <andrew@goteleport.com>

nklaassen committed:
use bytes.Join

nklaassen committed:
refactor ca pin checking loop

Co-authored-by: Andrew Lytvynov <andrew@goteleport.com>

nklaassen committed:
fix formatting

nklaassen committed:
use apiutils.Strings to allow scalar or list

nklaassen committed:
Merge branch 'master' into nklaassen/ca-pin-multiple-keys
